dune/grid/common/WellConnections.cpp

```diff
@@ -0,0 +1,197 @@
+        }
+        ++index;
+    }
+}
+
+std::vector<std::vector<int> >
+postProcessPartitioningForWells(std::vector<int>& parts,
+                                const Opm::EclipseStateConstPtr eclipseState,
+                                const WellConnections& well_connections,
+                                std::size_t no_procs)
+{
+    std::vector<const Opm::Well*>  wells  = eclipseState->getSchedule()->getWells();
+    // Contains for each process the indices of the wells assigned to it.
+    std::vector<std::vector<int> > well_indices_on_proc(no_procs);
+
+    if( ! well_connections.size() )
+    {
+        // No wells to be processed
+        return well_indices_on_proc;
+    }
+
+    // prevent memory allocation
+    for(auto& well_indices : well_indices_on_proc)
+    {
+        well_indices.reserve(wells.size());
+    }
+
+    // Check that all completions of a well have ended up on one process.
+    // If that is not the case for well then move them manually to the
+    // process that already has the most completions on it.
+    int well_index = 0;
+
+    for (const auto* well: wells) {
+        const auto& well_completions = well_connections[well_index];
+        std::map<int,std::size_t> no_completions_on_proc;
+        for ( auto completion_index: well_completions )
+        {
+            ++no_completions_on_proc[parts[completion_index]];
+        }
+
+        int owner = no_completions_on_proc.begin()->first;
+
+        if ( no_completions_on_proc.size() > 1 )
+        {
+            // partition with the most completions on it becomes new owner
+            int new_owner = std::max_element(no_completions_on_proc.begin(),
+                                             no_completions_on_proc.end(),
+                                             [](const std::pair<int,std::size_t>& p1,
+                                                const std::pair<int,std::size_t>& p2){
+                                                 return ( p1.second > p2.second );
+                                             })->first;
+            std::cout << "Manually moving well " << well->name() << " to partition "
+                      << new_owner << std::endl;
+
+            for ( size_t c = 0; c < well_completions.size(); c++ )
+            {
+                parts[c] = new_owner;
+            }
+
+            owner = new_owner;
+        }
+
+        well_indices_on_proc[owner].push_back(well_index);
+        ++well_index;
+    }
+    return well_indices_on_proc;
+
+}
+
+#ifdef HAVE_MPI
+std::unordered_set<std::string>
+computeDefunctWellNames(const std::vector<std::vector<int> >& wells_on_proc,
+                        const Opm::EclipseStateConstPtr eclipseState,
+                        const CollectiveCommunication<MPI_Comm>& cc,
+                        int root)
+{
+    std::vector<const Opm::Well*>  wells  = eclipseState->getSchedule()->getWells();
+    std::vector<int> my_well_indices;
+    const int well_information_tag = 267553;
```

Comment:
If this is required to be the same tag as in the other place it was used, the name must be in a header. Otherwise (it is independent) this is fine of course.